src/circuits/simple_gadgets.tcc

```diff
@@ -68,7 +69,19 @@ xor_constant_gadget<FieldT>::xor_constant_gadget(
 template<typename FieldT>
 void xor_constant_gadget<FieldT>::generate_r1cs_constraints()
 {
-    // 32 constraints
+    // Set the constraints (#constraints = length of bit string)
+    //
+    // We know that: res = a XOR b <=> (2.a) * b = a + b - res
+    // we can write: res = a + b - 2ab
+    //
+    // Hence, res2 = a XOR b XOR c = (a XOR b) XOR c <=> res XOR c
+    // Thus: res2 = res XOR c is constrainted as:
+    //   2(res) * c = res + c - res2
+    // which leads to:
+    //   2(a + b - 2ab) * c = a + b - 2ab + c - res2
+    // => res2 = 2ac + 2bc - 4abc - a - b + 2ab - c
+    // => res2 = b * [2 * (1 - 2c) * a] + b * (2c - 1) + a * (2c - 1) - c
+    // => res2 - b * (1 - 2c) - a * (1 - 2c) - c = b * [-2(1 - 2c) * a]
```

Comment:
There is a double negative error in line 84, we should have (after reordering a bit):
res2 + a*(1-2c) + b*(1-2c) + c = 2ab*(1-2c)

Reply:
Good catch thanks, hopefully it's fixed now